src/neovim.rs

```diff
@@ -108,13 +109,98 @@ where
     let (sender, receiver) = unbounded();
     let fut = future::try_join(
       req.clone().io_loop(reader, sender),
-      req.clone().handler_loop(handler, receiver)
+      req.clone().handler_loop(handler, receiver),
     )
     .map_ok(|_| ());
 
     (req, fut)
   }
 
+  /// Create a new instance, immediately send a handshake message and
+  /// wait for the response. Unlike `new`, this function is tolerant to extra
+  /// data in the reader before the handshake response is received.
+  pub async fn handshake<H, R>(
+    mut reader: R,
+    writer: W,
+    handler: H,
+  ) -> Result<
+    (
+      Neovim<<H as Handler>::Writer>,
+      impl Future<Output = Result<(), Box<LoopError>>>,
+    ),
+    Box<HandshakeError>,
+  >
+  where
+    R: AsyncRead + Send + Unpin + 'static,
+    H: Handler<Writer = W> + Spawner,
+  {
+    let instance = Neovim {
+      writer: Arc::new(Mutex::new(writer)),
+      msgid_counter: Arc::new(AtomicU64::new(0)),
+      queue: Arc::new(Mutex::new(Vec::new())),
+    };
+
+    let msgid = instance.msgid_counter.fetch_add(1, Ordering::SeqCst);
+    let magic = "nvim-rsMagicHandshakeMarkernvim-rs";
```

Comment:
I'm very hesitant to use a fixed string here. Please make that a function argument and push the responsibility of choosing it to the caller ;)

Reply:
Done